src/amp-react-selector.js

```diff
@@ -0,0 +1,142 @@
+ * limitations under the License.
+ */
+
+import ReactCompatibleBaseElement from './react-compat-base-element.js';
+import {
+  AmpContext,
+} from './amp-context.js';
+import {Slot} from './slot.js';
+
+const {
+  useContext,
+  useEffect,
+  useLayoutEffect,
+  useRef,
+  useState,
+} = preactHooks;
+
+
+export const AmpSelectorContext = preact.createContext();
+
+
+export function AmpSelector(props) {
+  const {multiple, value, defaultValue} = props;
+  const [selected, setSelectedState] = useState(
+    value ? [].concat(value) :
+    defaultValue ? [].concat(defaultValue) :
+    []);
+  const setSelectedRef = useRef(value => {
+    if (props.onChange) {
+      props.onChange({target: {value: multiple ? value : value[0]}});
+    }
+    setSelectedState(value);
+  });
+  return preact.createElement(
+    props.tagName || 'div',
+    {
+      ...props,
+    },
+    preact.createElement(
+      AmpSelectorContext.Provider,
+      {
+        ...props,
+        value: {
+          multiple,
+          // TBD: controlled values require override of properties.
+          selected: value ? [].concat(value) : selected,
+          setSelected: setSelectedRef.current,
+        },
+      }
+    )
+  );
+}
+
+AmpSelector.Option = function(props) {
+  const {option} = props;
+  const selectorContext = useContext(AmpSelectorContext);
+  const {selected, setSelected, multiple} = selectorContext;
+  const optionProps = {
+    ...props,
+    option,
+    selected: selected.includes(option),
+    onClick: () => selectOption(option, multiple, selected, setSelected),
+  };
+  if (props.render) {
+    return props.render(optionProps);
+  }
+  return preact.createElement(
+    props.type || props.tagName || 'div',
+    optionProps,
+    props.children);
+}
+
+function selectOption(option, multiple, selected, setSelected) {
+  if (multiple) {
+    setSelected(
+      selected.includes(option) ?
+      selected.filter(v => v != option) :
+      selected.concat(option));
+  } else {
+    if (!selected.includes(option)) {
+      setSelected([option]);
+    }
+  }
+}
+
+
+const AmpReactSelector = ReactCompatibleBaseElement(AmpSelector, {
+  className: 'i-amphtml-fill-content i-amphtml-replaced-content',
+  attrs: {
+  },
+  template: true,
+  // TODO: simply override ReactBaseElement instead.
+  init(reactBaseElement) {
+    const {element} = reactBaseElement;
+    const mu = new MutationObserver(records => {
+      rebuild();
+    });
+    const rebuild = () => {
+      const children = [];
+      const optionChildren = element.querySelectorAll('[option]');
+      const value = [];
+      optionChildren.forEach(child => {
+        // TODO: check that an option is not within another option.
+        const option = child.getAttribute('option');
+        const props = {
+          option,
+          type: Slot,
+          retarget: true,
+          assignedElements: [child],
+          postRender: () => {
```

Comment:
I think this is more of an AMP concern, instead of a React component one. How about if `init` returns an object that could have an optional `postRender` method, that way we can completely remove this from the React side?

Reply:
`Slot` component is managed by Preact, but otherwise it only exists for a single purpose of AMP/Preact integration. So it's not entirely too foreign here. But do you have any suggestion how we can do it better?

There are many things I don't like about this, but here's what I was trying to achieve. The main goal is to reduce render/re-render cycles. A mutation on a DOM element can lead to the component re-rendering (with new properties), which in turn can lead to DOM mutation. Typically this are benign and only cause extra re-rendering. But I did run into an infinite loop once or twice. Thus, we want to have a code such as:
```
pauseMutationObserver()
render()
resumeMutationObserver()
```

This poses a chain of its own problems:

1. Pause/resume observer is a bit expensive - it's essentially equivalent to disconnect/reconnect observer. It's cheaper to just call `observer.takeRecords()`. Not sure how much cheaper this is - we need to profile.
2. `observer.takeRecords()` could be dangerous - we could accidentally purge other mutation records that do not belong to `Slot` implementation, if they happen to have happened very close to each other. I used it here as a first draft, but this is likely not the last on this.
3. If we do adopt pause/resume - resume is harder to do. We have to repeat `observe()` with all the same configuration.
4. How do we actually pass the wrapper (or pause/resume pair) to the `Slot` component? The only options appear to be `props` or `context`. Since `Slot` is very AMP-specific, I decided to just use `props` instead of `context`.

As we develop it, we could make it look cleaner using some sort of `useAmpRenderingEffect` that would do pause/resume auto-magically. But we're not quite there yet.

Reply:
> A mutation on a DOM element can lead to the component re-rendering (with new properties), which in turn can lead to DOM mutation. Typically this are benign and only cause extra re-rendering. But I did run into an infinite loop once or twice.

What if we use a `passthrough` default slot for the light tree? Or is it a mutation on the light tree that's causing the infinite loop?

Reply:
For a selector, all mutations are in the main DOM tree and all observable by the main script/css/etc. E.g. a CSS like `.option[selected] {border: 1px solid red}`. That's currently virtually the only option for us to signify that a selection option (as a DOM element) is currently selected. So a subtree-based MutationObserver would see these mutations too. E.g. both a valid code paths:

/1/ User clicks on an option and the React component sets it as selected. Our `Slot` delegation updates DOM:
```
// A. React's onClick handler -> state
function AmpSelection.Option() {
  ...
  (<x onClick={() => setSelectedOptionState(props.option)} ...>)
  ...
}

// B. Slot's props.selected is set in React:
function AmpSelection.Option() {
  ...
  (<Slot selected={isSelectedOptionState(props.option)}>)
  ...
}

// C. Slot's side effect sets DOM attribute:

```
function Slot() {
  const domRef = useRef();
  useEffect(() => {
    const slot = domRef.current;
    ...
    // Update in the main DOM:
    const assignedOption = slot.assignedElements()[0];
    if (props.selected) {
      assignedOption.setAttribute('selected', '');
    } else {
      assignedOption.removeAttribute('selected');
    }
  });
  ...
}
```

/2/ A script in the main document manually writes DOM:

```
button1.onclick = () => {
  option2.setAttribute('selected', '');
};
```

We need the mutation observer to synchronize DOM -> React component in the case /2/. But we don't really need mutation observer for /1/ since we ourselves ensure that DOM/React are in full sync. In general case, incorrectly working /1/ can cause cycles. So far the cycles in such mutations have been easy to work around or ignore. But in general case this is still a dangerous situation. IMHO it'd be nice to have a more general solution for this.

And, finally, writing `selected` attribute is definitely not great. It'd be much better from API point of view if we could set a custom state, e.g. `.option:selected {...}`. But custom states spec is still ways and ways away.

Reply:
Filed https://github.com/ampproject/amp-react-prototype/issues/40 for this nuance.